### ISSUE TITLE
OpenCL CL_INVALID_DEVICE (-33) error in opencl_common.c:452 - Error querying PLATFORM_NAME

### ISSUE THREAD

**Issue by madtomic** [magnumripper/JohnTheRipper#3450]
**Environment Setup**

OS: Ubuntu 16.04.5
JohnTheRipper
https://github.com/magnumripper/JohnTheRipper/blob/bleeding-jumbo/doc/INSTALL-UBUNTU
***configure and make using this install guide.
**What was used:**
sudo apt-get install build-essential libssl-dev git zlib1g-dev
sudo apt-get install yasm libgmp-dev libpcap-dev pkg-config libbz2-dev
sudo apt-get install nvidia-opencl-dev
sudo apt-get install libopenmpi-dev openmpi-bin
./configure --enable-mpi
make -s clean && make -sj4

**Sample test file:**

./tezos2john.py 'put guide flat machine express cave hello connect stay local spike ski romance express brass' 'jbzbdybr.vpbdbxnn@tezos.example.org' 'tz1eTjPtwYjdcBMStwVdEcwY2YE3th1bXyMR' > tezos

**Running John**

tezos@tezos-Desktop:~/JohnTheRipper/run$ ./john --devices=gpu --fork=2 --format=tezos-opencl --session=tezos tezos
Using default input encoding: UTF-8
Loaded 1 password hash (tezos-opencl, Tezos Key [PBKDF2-SHA512 OpenCL])
Cost 1 (iteration count) is 2048 for all loaded hashes
Will run 4 OpenMP threads per process (8 total across 2 processes)
Node numbers 1-2 of 2 (fork)
**OpenCL CL_INVALID_DEVICE (-33) error in opencl_common.c:452 - Error querying PLATFORM_NAME**
Device 0@tezos-Desktop: GeForce GTX 1070 Ti
Press 'q' or Ctrl-C to abort, almost any other key for status
1 0g 0:00:00:04  3/3 0g/s 38678p/s 38678c/s 38678C/s GPU:61°C clmom..mia196
1 0g 0:00:00:09  3/3 0g/s 51592p/s 51592c/s 51592C/s GPU:61°C 196642..biliyou
1 0g 0:00:00:11  3/3 0g/s 54131p/s 54131c/s 54131C/s GPU:60°C 115la..0998am
Waiting for 1 child to terminate
Session aborted

**BUG/Issues:** It can't run across multiple GPUs

**System Information:**

tezos@tezos-Desktop:~/JohnTheRipper/run$ lscpu
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                8
On-line CPU(s) list:   0-7
Thread(s) per core:    2
Core(s) per socket:    4
Socket(s):             1
NUMA node(s):          1
Vendor ID:             AuthenticAMD
CPU family:            21
Model:                 2
Model name:            AMD FX(tm)-8350 Eight-Core Processor
Stepping:              0
CPU MHz:               1411.029
CPU max MHz:           4000.0000
CPU min MHz:           1400.0000
BogoMIPS:              8037.07
Virtualization:        AMD-V
L1d cache:             16K
L1i cache:             64K
L2 cache:              2048K
L3 cache:              8192K
NUMA node0 CPU(s):     0-7
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 popcnt aes xsave avx f16c lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs xop skinit wdt lwp fma4 tce nodeid_msr tbm topoext perfctr_core perfctr_nb cpb hw_pstate ssbd ibpb vmmcall bmi1 arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold

tezos@tezos-Desktop:~/JohnTheRipper/run$ ./john --list=build-info
Version: 1.8.0.13-jumbo-1-bleeding-9c715c6 2018-11-04 18:00:16 +0530
Build: linux-gnu 64-bit x86_64 XOP AC MPI + OMP
SIMD: XOP, interleaving: MD4:2 MD5:2 SHA1:1 SHA256:2 SHA512:1
CPU tests: XOP
$JOHN is ./
Format interface version: 14
Max. number of reported tunable costs: 4
Rec file version: REC4
Charset file version: CHR3
CHARSET_MIN: 1 (0x01)
CHARSET_MAX: 255 (0xff)
CHARSET_LENGTH: 24
SALT_HASH_SIZE: 1048576
Max. Markov mode level: 400
Max. Markov mode password length: 30
gcc version: 5.4.0
GNU libc version: 2.23 (loaded: 2.23)
OpenCL headers version: 2.0
Crypto library: OpenSSL
OpenSSL library version: 01000207f
OpenSSL 1.0.2g  1 Mar 2016
GMP library version: 6.1.0
File locking: fcntl()
fseek(): fseek
ftell(): ftell
fopen(): fopen
memmem(): System's

tezos@tezos-Desktop:~/JohnTheRipper/run$ ./john --list=opencl-devices
Platform #0 name: NVIDIA CUDA, version: OpenCL 1.2 CUDA 10.0.185

    Device #0 (0) name:     GeForce GTX 1070 Ti
    Device vendor:          NVIDIA Corporation
    Device type:            GPU (LE)
    Device version:         OpenCL 1.2 CUDA
    Driver version:         410.73 [recommended]
    Native vector widths:   char 1, short 1, int 1, long 1
    Preferred vector width: char 1, short 1, int 1, long 1
    Global Memory:          7.9 GB
    Global Memory Cache:    304 KB
    Local Memory:           48 KB (Local)
    Constant Buffer size:   64 KB
    Max memory alloc. size: 2 GB
    Max clock (MHz):        1683
    Profiling timer res.:   1000 ns
    Max Work Group Size:    1024
    Parallel compute cores: 19
    CUDA cores:             2432  (19 x 128)
    Speed index:            4093056
    Warp size:              32
    Max. GPRs/work-group:   65536
    Compute capability:     6.1 (sm_61)
    Kernel exec. timeout:   yes
    NVML id:                0
    PCI device topology:    01:00.0
    PCI lanes:              16/16
    Fan speed:              22%
    Temperature:            59°C
    Utilization:            4%

    Device #1 (1) name:     GeForce GTX 1070 Ti
    Device vendor:          NVIDIA Corporation
    Device type:            GPU (LE)
    Device version:         OpenCL 1.2 CUDA
    Driver version:         410.73 [recommended]
    Native vector widths:   char 1, short 1, int 1, long 1
    Preferred vector width: char 1, short 1, int 1, long 1
    Global Memory:          8 GB
    Global Memory Cache:    304 KB
    Local Memory:           48 KB (Local)
    Constant Buffer size:   64 KB
    Max memory alloc. size: 2 GB
    Max clock (MHz):        1683
    Profiling timer res.:   1000 ns
    Max Work Group Size:    1024
    Parallel compute cores: 19
    CUDA cores:             2432  (19 x 128)
    Speed index:            4093056
    Warp size:              32
    Max. GPRs/work-group:   65536
    Compute capability:     6.1 (sm_61)
    Kernel exec. timeout:   no
    NVML id:                1
    PCI device topology:    05:00.0
    PCI lanes:              4/16
    Fan speed:              0%
    Temperature:            34°C
    Utilization:            9%

**Comment by magnumripper** [comment]
https://www.openwall.com/lists/john-users/2018/11/06/2

Does it always fail like this? I have seen some strange race condition problem with later nvidia drivers and I've been testing a work-around hack for some time. Perhaps it's time to merge that to the repo.

You could always try running two sessions (in two different terminal sessions) without fork, like this:
```
./john -dev=0 -node=1/2 -form=tezos-opencl -ses=tezos1 tezos -inc
```
```
./john -dev=1 -node=2/2 -form=tezos-opencl -ses=tezos2 tezos -inc
```
Please test that and report whether it works or not.

**Comment by magnumripper** [comment]
FWIW, it works fine on my Macbook as well as on Openwall's test rig:
```
$ ../run/john --devices=6,7 --fork=2 --format=tezos-opencl --session=tezos tezos.in -w -ru
Using default input encoding: UTF-8
Loaded 1 password hash (tezos-opencl, Tezos Key [PBKDF2-SHA512 OpenCL])
Cost 1 (iteration count) is 2048 for all loaded hashes
Will run 16 OpenMP threads per process (32 total across 2 processes)
Node numbers 1-2 of 2 (fork)
Device 7: GeForce GTX TITAN
Device 6: GeForce GTX TITAN X
Each node loaded 1/2 of wordfile to memory (about 12 KB/node)
Press 'q' or Ctrl-C to abort, almost any other key for status
1 0g 0:00:00:01 DONE (2018-11-07 00:40) 0g/s 43217p/s 43217c/s 43217C/s GPU:44C 12345..Sssing
Waiting for 1 child to terminate
2 0g 0:00:00:03 DONE (2018-11-07 00:40) 0g/s 20395p/s 20395c/s 20395C/s GPU:45C 7network..Notuseding
Session completed
```

**Comment by claudioandre-br** [comment]
I remember this `(-33)` error. When that happened, it was not related to fork.  An interesting test is to run JtR in all devices (up to an "invalid device" error). For example:
```
./john --devices=0 --format=tezos-opencl --session=tezos0 --max-run=30 tezos
./john --devices=1 --format=tezos-opencl --session=tezos1 --max-run=30 tezos
./john --devices=2 --format=tezos-opencl --session=tezos2 --max-run=30 tezos
./john --devices=3 --format=tezos-opencl --session=tezos3 --max-run=30 tezos
```

To confirm it is related to fork (or not).

**Comment by madtomic** [comment]
@magnumripper 

>It always fails when I tried like this. I am using a slight older driver 384.130 now and I am still getting the same error: OpenCL CL_INVALID_DEVICE (-33) error in opencl_common.c:452 - Error querying PLATFORM_NAME

tezos@tezos-Desktop:~/JohnTheRipper/run$ ./john --devices=0,1 --fork=2 --format=tezos-opencl --session=tezos tezos -w -ru
Using default input encoding: UTF-8
Loaded 1 password hash (tezos-opencl, Tezos Key [PBKDF2-SHA512 OpenCL])
Cost 1 (iteration count) is 2048 for all loaded hashes
Will run 4 OpenMP threads per process (8 total across 2 processes)
Node numbers 1-2 of 2 (fork)
OpenCL CL_INVALID_DEVICE (-33) error in opencl_common.c:452 - Error querying PLATFORM_NAME
Device 0@tezos-Desktop: GeForce GTX 1070 Ti
Each node loaded 1/2 of wordfile to memory (about 12 KB/node)
Press 'q' or Ctrl-C to abort, almost any other key for status
1 0g 0:00:00:01 DONE (2018-11-06 20:36) 0g/s 45227p/s 45227c/s 45227C/s GPU:58°C 12345..Sssing
Waiting for 1 child to terminate
Main process session completed, but some child processes failed


>Running in two different terminal sessions without fork seems to work but this raise another question when doing it this way. **Question:** Would this be doing the same work twice if doing it this way?

>Eventually, the end game setup is running this across 6-12 GeForce GTX 1070 Ti GPUs with AMD Threadripper 2950x as the CPU. I am only doing the preliminary system tests before I commit to this project. 

tezos@tezos-Desktop:~/JohnTheRipper/run$ OMP_NUM_THREADS=4 ./john -dev=0 -node=1/2 -form=tezos-opencl -ses=tezos1 tezos -inc
Device 0@tezos-Desktop: GeForce GTX 1070 Ti
Using default input encoding: UTF-8
Loaded 1 password hash (tezos-opencl, Tezos Key [PBKDF2-SHA512 OpenCL])
Cost 1 (iteration count) is 2048 for all loaded hashes
Will run 4 OpenMP threads
Node number 1 of 2
Press 'q' or Ctrl-C to abort, almost any other key for status
0g 0:00:00:55  0g/s 57326p/s 57326c/s 57326C/s GPU:65°C sulsio..amr3mk
0g 0:00:01:26  0g/s 57634p/s 57634c/s 57634C/s GPU:66°C phyrero..178jkl
0g 0:00:01:54  0g/s 57970p/s 57970c/s 57970C/s GPU:66°C sanno181..pandysum
0g 0:00:03:15  0g/s 58256p/s 58256c/s 58256C/s GPU:67°C legl7w..dm120*
Session aborted

tezos@tezos-Desktop:~/JohnTheRipper/run$ OMP_NUM_THREADS=4 ./john -dev=1 -node=2/2 -form=tezos-opencl -ses=tezos2 tezos -inc
Device 1@tezos-Desktop: GeForce GTX 1070 Ti
Using default input encoding: UTF-8
Loaded 1 password hash (tezos-opencl, Tezos Key [PBKDF2-SHA512 OpenCL])
Cost 1 (iteration count) is 2048 for all loaded hashes
Will run 4 OpenMP threads
Node number 2 of 2
Press 'q' or Ctrl-C to abort, almost any other key for status
0g 0:00:00:40  0g/s 56295p/s 56295c/s 56295C/s GPU:53°C 019m5n..chybam
0g 0:00:01:12  0g/s 57255p/s 57255c/s 57255C/s GPU:57°C abi00..tt29c
0g 0:00:01:42  0g/s 57963p/s 57963c/s 57963C/s GPU:60°C muradogi..cliser22
0g 0:00:02:57  0g/s 58405p/s 58405c/s 58405C/s GPU:64°C sucefass..prylos10
Session aborted

**Comment by madtomic** [comment]
@claudioandre-br 

>Interesting test

tezos@tezos-Desktop:~/JohnTheRipper/run$ ./john --devices=0 --format=tezos-opencl --session=tezos0 --max-run=30 tezos
Device 0@tezos-Desktop: GeForce GTX 1070 Ti
Using default input encoding: UTF-8
Loaded 1 password hash (tezos-opencl, Tezos Key [PBKDF2-SHA512 OpenCL])
Cost 1 (iteration count) is 2048 for all loaded hashes
Will run 8 OpenMP threads
Press 'q' or Ctrl-C to abort, almost any other key for status
0g 0:00:00:30  3/3 0g/s 51864p/s 51864c/s 51864C/s GPU:65°C hyelu..ccmcsc
0g 0:00:00:30  3/3 0g/s 53041p/s 53041c/s 53041C/s GPU:62°C ccmcso..02428a
Session stopped (max run-time reached)

tezos@tezos-Desktop:~/JohnTheRipper/run$ ./john --devices=1 --format=tezos-opencl --session=tezos1 --max-run=30 tezos
Device 1@tezos-Desktop: GeForce GTX 1070 Ti
Using default input encoding: UTF-8
Loaded 1 password hash (tezos-opencl, Tezos Key [PBKDF2-SHA512 OpenCL])
Cost 1 (iteration count) is 2048 for all loaded hashes
Will run 8 OpenMP threads
Press 'q' or Ctrl-C to abort, almost any other key for status
0g 0:00:00:26  3/3 0g/s 50845p/s 50845c/s 50845C/s GPU:54°C 067644..rjbbs
0g 0:00:00:30  3/3 0g/s 54363p/s 54363c/s 54363C/s GPU:52°C ccmcso..02428a
Session stopped (max run-time reached)

**Comment by madtomic** [comment]
@magnumripper 

I decided to let it run awhile and it would randomly abort the session if it is in separate terminal session. So this is would not a viable solution.

tezos@tezos-Desktop:~/JohnTheRipper/run$ OMP_NUM_THREADS=4 ./john -dev=0 -node=1/2 -form=tezos-opencl -ses=tezos1 tezos -inc
Device 0@tezos-Desktop: GeForce GTX 1070 Ti
Using default input encoding: UTF-8
Loaded 1 password hash (tezos-opencl, Tezos Key [PBKDF2-SHA512 OpenCL])
Cost 1 (iteration count) is 2048 for all loaded hashes
Will run 4 OpenMP threads
Node number 1 of 2
Press 'q' or Ctrl-C to abort, almost any other key for status
0g 0:00:00:34  0g/s 54886p/s 54886c/s 54886C/s GPU:61°C ppawei..mampse10
0g 0:00:01:21  0g/s 56658p/s 56658c/s 56658C/s GPU:66°C sustsot1..chuchk06
0g 0:00:02:38  0g/s 57349p/s 57349c/s 57349C/s GPU:67°C lyoni03..rogdis2
Session aborted

**Comment by claudioandre-br** [comment]
I said something wrong above. The `-33` is related to `--fork`.
- https://github.com/magnumripper/JohnTheRipper/issues/2547#issuecomment-322013050

As I wrote there, the error happens in some machines: yours is the 2nd. But, for example, it works in your Ubuntu 14 and our super machine. I still can't see why. 

BTW:
> Question: Would this be doing the same work twice if doing it this way?

No.

**Comment by kholia** [comment]
> Eventually, the end game setup is running this across 6-12 GeForce GTX 1070 Ti GPUs with AMD Threadripper 2950x as the CPU.

> I did noticed an increase in performance when using a dual core cpu vs quad cpu.

This format is not a pure GPU format. It uses the CPUs heavily for doing asymmetric crypto stuff.

I am guessing that a single Threadripper 2950x CPU won't be enough when working with 6-12 fast GPUs.

Perhaps, the workload can be divided among multiple computers using something like OpenMP? CC @magnumripper.

**Comment by madtomic** [comment]
@kholia 
>This format is not a pure GPU format. It uses the CPUs heavily for doing asymmetric crypto stuff.
>I am guessing that a single Threadripper 2950x CPU won't be enough when working with 6-12 fast GPUs.

Threadripper 2950x CPU has over 32 threads. If I pair up with just 6 GPUs. I would allocated 5-6 threads per GPU.

I think 2970X would do a better job as it has 48 threads or 8 OpenMP threads per card when pair up with 6 cards?

**Comment by claudioandre-br** [comment]
> It seems like I am having the very same issue a year ago. I will try Ubuntu 14 with the same Nvidia cards to see it it would works like AMD cards.

If you can, please, do some trials in order to narrow the problem. In fact, since it works on super, I would start trying CUDA 7 (using whatever Ubuntu version you are happy with).

> I decided to let it run awhile and it would randomly abort the session if it is in separate terminal session. So this is would not a viable solution.

This is scaring. It means you can't use 2 GPUs at the same time. Even in different and unrealated processes!?! Odd.

**Comment by claudioandre-br** [comment]
And please, run other tests. E.g.:

* 1 GPU, two processes
```
./john -form:cpu --list=format-tests 2> /dev/null | cut -f3 1> alltests.in
./john --fork=2 --devices=0 --format=raw-sha256-opencl alltests.in
```
* And, using separate terminals JtR randomly aborts the session
```
./john --devices=0 --format=raw-sha256-opencl alltests.in -session=1 # terminal 1
./john --devices=1 --format=raw-sha512-opencl alltests.in -session=2 # terminal 2
```
*****
I'm running these tests now on super (on -dev 6 and 7). 
On super everything seems ok. And it has CUDA 8 now (but it is hybrid AMD/NVIDIA env).
```
Platform #2 name: NVIDIA CUDA, version: OpenCL 1.2 CUDA 8.0.13
    Device version:         OpenCL 1.2 CUDA
    Driver version:         361.28 [recommended]
```

**Comment by madtomic** [comment]
@claudioandre-br 

>If you can, please, do some trials in order to narrow the problem. In fact, since it works on super, I would start trying CUDA 7 (using whatever Ubuntu version you are happy with).

After few hours of trying. I have given up on this trial. Ubuntu 14 does not like the new NVIDIA cards. I guess, something in the NVIDIA driver that Ubuntu 14 doesn't like. I will keep trying with older drivers. I am sure there has to be a working driver for Ubuntu 14 for the 1070 ti cards.

>This is scaring. It means you can't use 2 GPUs at the same time. Even in different and unrealated processes!?! Odd.

I found out that I have aborted the session myself. No issue here running in separate terminals.

**Comment by madtomic** [comment]
@claudioandre-br

tezos@tezos-MS-7577:~/JohnTheRipper/run$ ./john -form:cpu --list=format-tests 2> /dev/null | cut -f3 1> alltests.in
tezos@tezos-MS-7577:~/JohnTheRipper/run$ ./john --fork=2 --devices=0 --format=raw-sha256-opencl alltests.in
Using default input encoding: UTF-8
Loaded 87 password hashes with no different salts (raw-SHA256-opencl [SHA256 OpenCL])
Remaining 65 password hashes with no different salts
Node numbers 1-2 of 2 (fork)
OpenCL CL_INVALID_DEVICE (-33) error in opencl_common.c:452 - Error querying PLATFORM_NAME
Device 0@tezos-MS-7577: GeForce GTX 1070 Ti
Note: This format may be a lot faster with --mask acceleration (see doc/MASK).
Press 'q' or Ctrl-C to abort, almost any other key for status
1 0g 0:00:01:18  3/3 0g/s 36764Kp/s 36764Kc/s 2397MC/s GPU:59°C myjauei..mnneitl

tezos@tezos-MS-7577:~/JohnTheRipper/run$ ./john --devices=0 --format=raw-sha256-opencl alltests.in -session=1 # terminal 1
Device 0@tezos-MS-7577: GeForce GTX 1070 Ti
Using default input encoding: UTF-8
Loaded 87 password hashes with no different salts (raw-SHA256-opencl [SHA256 OpenCL])
Remaining 65 password hashes with no different salts
Note: This format may be a lot faster with --mask acceleration (see doc/MASK).
Press 'q' or Ctrl-C to abort, almost any other key for status
11               (?)
11               (?)
2                (?)
2                (?)
4g 0:00:01:22  3/3 0.04870g/s 32143Kp/s 32143Kc/s 1963MC/s GPU:62°C revdlia..rb15p55

tezos@tezos-MS-7577:~/JohnTheRipper/run$ ./john --devices=1 --format=raw-sha512-opencl alltests.in -session=2 # terminal 2
Device 1@tezos-MS-7577: GeForce GTX 1070 Ti
Using default input encoding: UTF-8
Loaded 35 password hashes with no different salts (raw-SHA512-opencl [SHA512 OpenCL])
Note: This format may be a lot faster with --mask acceleration (see doc/MASK).
Press 'q' or Ctrl-C to abort, almost any other key for status
                 (?)
                 (?)
password         (?)
password         (?)
12345678         (?)
12345678         (?)
12345678         (?)
12345678         (?)
8g 0:00:00:46  3/3 0.1711g/s 23079Kp/s 23079Kc/s 633051KC/s GPU:46°C fsbu1aa..napug0$

**Comment by madtomic** [comment]
@magnumripper

I gotten this when trying this.

tezos@tezos-MS-7577:~/JohnTheRipper/run$ mpirun -np 2 ./john --devices=gpu --format=tezos-opencl tezos
Device 1@tezos-MS-7577: GeForce GTX 1070 Ti
Device 0@tezos-MS-7577: GeForce GTX 1070 Ti
Using default input encoding: UTF-8
Loaded 1 password hash (tezos-opencl, Tezos Key [PBKDF2-SHA512 OpenCL])
Cost 1 (iteration count) is 2048 for all loaded hashes
MPI in use, disabling OMP (see doc/README.mpi)
Node numbers 1-2 of 2 (MPI)
Send SIGUSR1 to mpirun for status

It looks like it crashed?

**Comment by magnumripper** [comment]
> It looks like it crashed?

From that output there's no sign it crashed. Like the output says, you can't press a key to get status but have to send s SIGUSR1 to each john process or their parent mpirun process to get a status output.

**Comment by magnumripper** [comment]
> `MPI in use, disabling OMP (see doc/README.mpi)`

To get eg. 4 threads per process, use `mpirun -np 2 -x OMP_NUM_THREADS=4`

**Comment by madtomic** [comment]
This is what I have gather so far, more threads per process has an impact on password guesses per second. It seems to me that running a single card with most threads would be the ideal set-up to get the most out of the card and it would use less electric as well.

**I wish, I have 16 threads to test these 2 cards and run 8 threads per card and see if I would get 70K+ p/s per card.** Can someone please simulate this test for me?

Technical question. I read that MPI and OpenMP can be used together. Would this bring better performance?

**Single card method**
This method using 8 threads per process yields about 70K+ p/s per card. **Total 70K+ p/s**
./john --devices=0 --format=tezos-opencl --session=tezos1 tezos

**Split terminals method - 2 devices**
This method using 4 threads per process yields about 60K+ p/s per card. **Total 120K+ p/s**
OMP_NUM_THREADS=4 ./john -dev=0 -node=1/2 -form=tezos-opencl -ses=tezos1 tezos -inc
OMP_NUM_THREADS=4 ./john -dev=1 -node=2/2 -form=tezos-opencl -ses=tezos2 tezos -inc

**mpirun method - 2 devices**
This method using 4 threads per process yields about 42K+ p/s per card. **Total 84K+ p/s**
mpirun -np 2 -x OMP_NUM_THREADS=4 ./john --devices=gpu --format=tezos-opencl --progress-every=10 --session=tezos tezos

Didn't work:
./john --devices=gpu --fork=2 --format=tezos-opencl --session=tezos tezos
I'm am still wondering what the password guesses would be for this method.

**Comment by magnumripper** [comment]
> I wish, I have 16 threads to test these 2 cards and run 8 threads per card and see if I would get 70K+ p/s per card. Can someone please simulate this test for me?

If you get 70K+ using one card and 8 threads, you will certainly get 70K+ *per* card with *n* cards as long as you have 8 threads per card. The question is whether 8 threads per cards is enough.

Here's Openwall's test rig, with a single Titan X:

4 threads:
```
[magnum@super src]$ OMP_NUM_THREADS=4 ../run/john -form:tezos-opencl -test
../run/john: /usr/local/cuda-8.0/targets/x86_64-linux/lib/libOpenCL.so.1: no version information available (required by ../run/john)
Will run 4 OpenMP threads
Device 6: GeForce GTX TITAN X
Benchmarking: tezos-opencl, Tezos Key [PBKDF2-SHA512 OpenCL]... (4xOMP) DONE
Speed for cost 1 (iteration count) of 2048
Raw:	58864 c/s real, 29970 c/s virtual
```

8 threads:
```
[magnum@super src]$ OMP_NUM_THREADS=8 ../run/john -form:tezos-opencl -test
../run/john: /usr/local/cuda-8.0/targets/x86_64-linux/lib/libOpenCL.so.1: no version information available (required by ../run/john)
Will run 8 OpenMP threads
Device 6: GeForce GTX TITAN X
Benchmarking: tezos-opencl, Tezos Key [PBKDF2-SHA512 OpenCL]... (8xOMP) DONE
Speed for cost 1 (iteration count) of 2048
Raw:	70217 c/s real, 29257 c/s virtual
```

16 threads:
```
$ OMP_NUM_THREADS=16 ../run/john -form:tezos-opencl -test
../run/john: /usr/local/cuda-8.0/targets/x86_64-linux/lib/libOpenCL.so.1: no version information available (required by ../run/john)
Will run 16 OpenMP threads
Device 6: GeForce GTX TITAN X
Benchmarking: tezos-opencl, Tezos Key [PBKDF2-SHA512 OpenCL]... (16xOMP) DONE
Speed for cost 1 (iteration count) of 2048
Raw:	77404 c/s real, 28248 c/s virtual, GPU util: 91%
```

> Technical question. I read that MPI and OpenMP can be used together. Would this bring better performance?

No, fork and OpenMP is also used together: When you say `--fork=2` the number of threads per process automagically halves so we don't overbook the CPU. In the MPI case, we default to single thread per process because we don't have enough information - but if you pass `-x OMP_NUM_THREADS` to `mpirun` we use that.

I'm not sure why you get 120K using `-node` and just 84K using MPI. You should **really** get the same speed for those two cases - technically they are just about the same thing.

I have pushed some of my race-condition workarounds to a separate branch `nvidia-raceco`. Try checking that out and see if it helps for the `--fork` case and please report back!

**Comment by madtomic** [comment]
@magnumripper

Thank you for doing the simulation tests.

Using nvidia-raceco branch with --fork option, I still get the same error.
OpenCL CL_INVALID_DEVICE (-33) error in opencl_common.c:453 - Error querying  PLATFORM_NAME

Is there something else I should be trying using this branch?

**Comment by magnumripper** [comment]
I pushed another commit to the nvidia-raceco branch: It will now sleep for a split second and retry up to 10 times. Please try that.

**Comment by claudioandre-br** [comment]
I would like to know what happens if you ignore the error. Like this:

```diff
diff --git a/src/opencl_common.c b/src/opencl_common.c
index 5d4efdc8c..f904c6473 100644
--- a/src/opencl_common.c
+++ b/src/opencl_common.c
@@ -446,14 +446,13 @@ static void start_opencl_environment()
 	clGetPlatformIDs(MAX_PLATFORMS, platform_list, &num_platforms);
 
 	for (i = 0; i < num_platforms; i++) {
-		platforms[i].platform = platform_list[i];
-
-		HANDLE_CLERROR(clGetPlatformInfo(platforms[i].platform,
+printf("Platform #%d %lu\n", i, platform_list[i]);
+		SOFT_CLERROR(clGetPlatformInfo(platform_list[i],
 		                                 CL_PLATFORM_NAME, sizeof(opencl_data), opencl_data, NULL),
 		               "Error querying PLATFORM_NAME");
 
 		// It is possible to have a platform without any devices
-		ret = clGetDeviceIDs(platforms[i].platform, CL_DEVICE_TYPE_ALL,
+		ret = clGetDeviceIDs(platform_list[i], CL_DEVICE_TYPE_ALL,
 		                     MAX_GPU_DEVICES, &devices[device_pos],
 		                     &device_num);
 
@@ -462,6 +461,7 @@ static void start_opencl_environment()
 			fprintf(stderr, "No OpenCL devices was found on platform #%d\n", i);
 
 		// Save platform and devices information
+		platforms[i].platform = platform_list[i];
 		platforms[i].num_devices = device_num;
 
 		// Point to the end of the list
```

**Comment by claudioandre-br** [comment]
And try to find new messages using increased verbosity
```
. /john --devices=gpu --fork=2 --format=tezos-opencl --session=tezos tezos -verb=5
```

**Comment by magnumripper** [comment]
> An explanation for what is happening is (...)

Yeah that's a plausible explanation. With current code (as of 77be5eb) we'd handle it better and if using `--verb:5` we'll also see a `No OpenCL devices found on platform #0` from that process.

If this is what's happening, we still need to work around some race condition in the driver so (1) doesn't happen in the first place: Perhaps we need a `do ... while (++retry < 10 && ret != CL_SUCCESS)` around the very first `clGetPlatformIDs` in L445 (old line number) and eventually make that thing silent (at least unless RACE_CONDITION_DEBUG) and non-error (for actual intended behavior when we don't *have* a platform).

**Comment by madtomic** [comment]
Clone the latest nvidia-raceco branch, commit: faf3175 - I receive this build error while make

tezos@tezos-MS-7577:~/Downloads/john/src$ make -s clean && make -sj4
ar: creating aes.a
ar: creating ed25519-donna.a
ar: creating secp256k1.a
stribog_fmt_plug.c:469:2: warning: #warning Stribog-256 and Stribog-512 formats require SSE 4.1, formats disabled [-Wcpp]
 #warning Stribog-256 and Stribog-512 formats require SSE 4.1, formats disabled
  ^
In file included from opencl_7z_fmt_plug.c:34:0:
opencl_7z_fmt_plug.c: In function ‘crypt_all’:
options.h:193:15: error: ‘mpi_p’ undeclared (first use in this function)
 #define NODE (mpi_p > 1 ? mpi_id + 1 : options.node_min)
               ^
opencl_common.h:241:6: note: in expansion of macro ‘NODE’
      NODE, get_error_name(__err), __FILE__, __LINE__, message); \
      ^
opencl_7z_fmt_plug.c:667:3: note: in expansion of macro ‘BENCH_CLERROR’
   BENCH_CLERROR(clEnqueueWriteBuffer(queue[gpu_id], mem_in, CL_FALSE, 0,
   ^
options.h:193:15: note: each undeclared identifier is reported only once for each function it appears in
 #define NODE (mpi_p > 1 ? mpi_id + 1 : options.node_min)
               ^
opencl_common.h:241:6: note: in expansion of macro ‘NODE’
      NODE, get_error_name(__err), __FILE__, __LINE__, message); \
      ^
opencl_7z_fmt_plug.c:667:3: note: in expansion of macro ‘BENCH_CLERROR’
   BENCH_CLERROR(clEnqueueWriteBuffer(queue[gpu_id], mem_in, CL_FALSE, 0,
   ^
options.h:193:27: error: ‘mpi_id’ undeclared (first use in this function)
 #define NODE (mpi_p > 1 ? mpi_id + 1 : options.node_min)
                           ^
opencl_common.h:241:6: note: in expansion of macro ‘NODE’
      NODE, get_error_name(__err), __FILE__, __LINE__, message); \
      ^
opencl_7z_fmt_plug.c:667:3: note: in expansion of macro ‘BENCH_CLERROR’
   BENCH_CLERROR(clEnqueueWriteBuffer(queue[gpu_id], mem_in, CL_FALSE, 0,
   ^
make[1]: *** [opencl_7z_fmt_plug.o] Error 1
make[1]: *** Waiting for unfinished jobs....
make: *** [default] Error 2

**Comment by madtomic** [comment]
@claudioandre-br 

Added your changes to opencl_common.c and here is my result.

tezos@tezos-MS-7577:~/JohnTheRipper/run$ ./john --devices=gpu --fork=2 --format=tezos-opencl --session=tezos tezos -verb=5
initUnicode(UNICODE, UTF-8/ISO-8859-1)
UTF-8 -> UTF-8 -> UTF-8
Using default input encoding: UTF-8
Loaded 1 password hash (tezos-opencl, Tezos Key [PBKDF2-SHA512 OpenCL])
Cost 1 (iteration count) is 2048 for all loaded hashes
Will run 2 OpenMP threads per process (4 total across 2 processes)
Node numbers 1-2 of 2 (fork)
OpenCL CL_INVALID_DEVICE (-33) error in opencl_common.c:452 - Error querying PLATFORM_NAME
Device 0@tezos-MS-7577: GeForce GTX 1070 Ti
Loaded 2 hashes with 2 different salts to test db from test vectors
Options used: -I /home/tezos/JohnTheRipper/run/kernels -cl-mad-enable -DSM_MAJOR=6 -DSM_MINOR=1 -cl-nv-verbose -D__GPU__ -DDEVICE_INFO=524306 -DSIZEOF_SIZE_T=8 -DDEV_VER_MAJOR=410 -DDEV_VER_MINOR=73 -D_OPENCL_COMPILER -DHASH_LOOPS=250 -DPLAINTEXT_LENGTH=110 -DPBKDF2_64_MAX_SALT_SIZE=107 $JOHN/kernels/tezos_kernel.cl
Build log: 
ptxas info    : 0 bytes gmem, 936 bytes cmem[3]
ptxas info    : Compiling entry function 'pbkdf2_sha512_kernel_varying_salt' for 'sm_61'
ptxas info    : Function properties for pbkdf2_sha512_kernel_varying_salt
ptxas         .     416 bytes stack frame, 0 bytes spill stores, 0 bytes spill loads
ptxas info    : Used 104 registers, 344 bytes cmem[0], 32 bytes cmem[2]
ptxas info    : Compiling entry function 'pbkdf2_sha512_kernel' for 'sm_61'
ptxas info    : Function properties for pbkdf2_sha512_kernel
ptxas         .     128 bytes stack frame, 0 bytes spill stores, 0 bytes spill loads
ptxas info    : Used 96 registers, 344 bytes cmem[0], 32 bytes cmem[2]
ptxas info    : Compiling entry function 'pbkdf2_sha512_loop' for 'sm_61'
ptxas info    : Function properties for pbkdf2_sha512_loop
ptxas         .     0 bytes stack frame, 0 bytes spill stores, 0 bytes spill loads
ptxas info    : Used 157 registers, 336 bytes cmem[0], 8 bytes cmem[2]
Calculating best GWS for LWS=32; max. 100ms single kernel invocation.
Raw speed figures including buffer transfers:
xfer: 4.096us, init: 98.304us, crypt: 8x9.010ms, res xfer: 3.456us
gws:       256	   3463c/s     7092224 rounds/s  73.917ms per crypt_all()!
xfer: 9.216us, init: 160.736us, crypt: 8x9.133ms, res xfer: 5.856us
gws:       512	   6827c/s    13981696 rounds/s  74.993ms per crypt_all()+
xfer: 23.712us, init: 160.768us, crypt: 8x8.975ms, res xfer: 10.400us
gws:      1024	  13890c/s    28446720 rounds/s  73.721ms per crypt_all()!
xfer: 45.728us, init: 160.768us, crypt: 8x8.396ms, res xfer: 20.352us
gws:      2048	  29675c/s    60774400 rounds/s  69.013ms per crypt_all()!
xfer: 89.696us, init: 168.960us, crypt: 8x8.420ms, res xfer: 40.096us
gws:      4096	  59123c/s   121083904 rounds/s  69.278ms per crypt_all()+
xfer: 178.112us, init: 224.096us, crypt: 8x17.421ms, res xfer: 79.968us
gws:      8192	  57207c/s   117159936 rounds/s 143.197ms per crypt_all()
xfer: 606.976us, init: 453.632us, crypt: 8x25.970ms, res xfer: 158.880us
gws:     16384	  76570c/s   156815360 rounds/s 213.971ms per crypt_all()+
xfer: 1.640ms, init: 940.032us, crypt: 8x42.652ms, res xfer: 321.792us
gws:     32768	  93008c/s   190480384 rounds/s 352.312ms per crypt_all()+
xfer: 4.212ms, init: 1.705ms, crypt: 8x70.558ms, res xfer: 938.208us
gws:     65536	 112051c/s   229480448 rounds/s 584.873ms per crypt_all()+
xfer: 9.343ms, init: 3.422ms, crypt: 8x134.391ms (exceeds 100ms)
Calculating best LWS for GWS=65536
Testing LWS=32 GWS=65536 ... 198.504ms+
Testing LWS=64 GWS=65536 ... 200.186ms
Testing LWS=128 GWS=65536 ... 200.629ms
Testing LWS=256 GWS=65536 ... 307.239ms
Calculating best GWS for LWS=32; max. 200ms single kernel invocation.
Raw speed figures including buffer transfers:
xfer: 53.952us, init: 146.304us, crypt: 8x7.218ms, res xfer: 24.544us
gws:      2432	  40973c/s    83912704 rounds/s  59.356ms per crypt_all()!
xfer: 106.400us, init: 160.640us, crypt: 8x7.961ms, res xfer: 47.392us
gws:      4864	  74218c/s   151998464 rounds/s  65.535ms per crypt_all()+
xfer: 214.016us, init: 249.856us, crypt: 8x15.653ms, res xfer: 94.720us
gws:      9728	  75530c/s   154685440 rounds/s 128.795ms per crypt_all()+
xfer: 508.960us, init: 489.472us, crypt: 8x21.879ms, res xfer: 204.064us
gws:     19456	 107824c/s   220823552 rounds/s 180.441ms per crypt_all()+
xfer: 1.269ms, init: 976.704us, crypt: 8x44.600ms, res xfer: 602.272us
gws:     38912	 105677c/s   216426496 rounds/s 368.214ms per crypt_all()
xfer: 3.237ms, init: 1.961ms, crypt: 8x79.862ms, res xfer: 1.190ms
gws:     77824	 117803c/s   241260544 rounds/s 660.625ms per crypt_all()+
xfer: 10.259ms, init: 3.908ms, crypt: 8x159.772ms, res xfer: 2.604ms
gws:    155648	 117414c/s   240463872 rounds/s    1.325s per crypt_all()
xfer: 16.652ms, init: 7.857ms, crypt: 8x311.184ms (exceeds 200ms)
Local worksize (LWS) 32, global worksize (GWS) 77824
Press 'q' or Ctrl-C to abort, almost any other key for status
1 0g 0:00:00:04  3/3 0g/s 19366p/s 19366c/s 19366C/s GPU:43°C util:100% fan:0% 123456
1 0g 0:00:00:07  3/3 0g/s 31642p/s 31642c/s 31642C/s GPU:40°C fan:0% mia199..koah2
1 0g 0:00:00:09  3/3 0g/s 33761p/s 33761c/s 33761C/s GPU:41°C fan:0% koahz..jmoutd
Waiting for 1 child to terminate
Session aborted

**Comment by claudioandre-br** [comment]
> Added your changes to opencl_common.c and here is my result.
*I did rebulid john after making changes. If you are wondering.
./configure --enable-mpi
make -s clean && make -sj4

That is strange. You added 5 new lines and it still is complaining about line 452?
>OpenCL CL_INVALID_DEVICE (-33) error in opencl_common.c:452 - Error querying PLATFORM_NAME
Device 0@tezos-MS-7577: GeForce GTX 1070 Ti

Anyway, if correct, it means `clGetPlatformIDs` works and `clGetPlatformInfo` fails (and this is good news). Well, kind of.

**Comment by magnumripper** [comment]
Most of the debug stuff are now in the main branch. To enable it, do this:
```diff
diff --git a/src/opencl_common.c b/src/opencl_common.c
index d7e210067..32d518a9b 100644
--- a/src/opencl_common.c
+++ b/src/opencl_common.c
@@ -63,7 +63,7 @@
 #define KLUDGE_LOCK_FILE "/tmp/.JtR_kernel_build_lock"
 
 /* Set this to eg. 3 for some added debug and retry stuff */
-#define RACE_CONDITION_DEBUG 0
+#define RACE_CONDITION_DEBUG 3
 
 #define LOG_SIZE 1024*16
 
```
...and then run with `--verb=5`.

**Comment by madtomic** [comment]
@magnumripper 

>I am using a new system and same two Nvidia video cards.
>I am still getting the same error, please see end for --verb=5

Node numbers 1-2 of 2 (fork)
No OpenCL devices were found on platform #0: CL_INVALID_DEVICE (-33) (-33)
No OpenCL devices found
Device 0: GeForce GTX 1070 Ti



tezos@tezos-MS-7577:~/john/JohnTheRipper/run$ lscpu
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                16
On-line CPU(s) list:   0-15
Thread(s) per core:    2
Core(s) per socket:    8
Socket(s):             1
NUMA node(s):          1
Vendor ID:             AuthenticAMD
CPU family:            23
Model:                 8
Stepping:              2
CPU MHz:               2200.000
BogoMIPS:              7400.00
Virtualization:        AMD-V
L1d cache:             32K
L1i cache:             64K
L2 cache:              512K
L3 cache:              8192K
NUMA node0 CPU(s):     0-15



tezos@tezos-MS-7577:~/john/JohnTheRipper/run$ ./john --list=build-info
Version: 1.8.0.13-jumbo-1-bleeding-43ed1a4 2018-11-23 19:32:37 +0100
Build: linux-gnu 64-bit x86_64 AVX2 AC OMP
SIMD: AVX2, interleaving: MD4:3 MD5:3 SHA1:1 SHA256:1 SHA512:1
CPU tests: AVX2
$JOHN is ./
Format interface version: 14
Max. number of reported tunable costs: 4
Rec file version: REC4
Charset file version: CHR3
CHARSET_MIN: 1 (0x01)
CHARSET_MAX: 255 (0xff)
CHARSET_LENGTH: 24
SALT_HASH_SIZE: 1048576
Max. Markov mode level: 400
Max. Markov mode password length: 30
gcc version: 4.8.4
GNU libc version: 2.19 (loaded: 2.19)
OpenCL headers version: 1.2
Crypto library: OpenSSL
OpenSSL library version: 01000106f
OpenSSL 1.0.1f 6 Jan 2014
GMP library version: 5.1.3
File locking: fcntl()
fseek(): fseek
ftell(): ftell
fopen(): fopen
memmem(): System's



tezos@tezos-MS-7577:~/john/JohnTheRipper/run$ ./john --devices=gpu --fork=2 --format=tezos-opencl --session=tezos tezos -inc --verb=5
initUnicode(UNICODE, UTF-8/ISO-8859-1)
UTF-8 -> UTF-8 -> UTF-8
Using default input encoding: UTF-8
Loaded 1 password hash (tezos-opencl, Tezos Key [PBKDF2-SHA512 OpenCL])
Cost 1 (iteration count) is 2048 for all loaded hashes
Will run 8 OpenMP threads per process (16 total across 2 processes)
Node numbers 1-2 of 2 (fork)
No OpenCL devices were found on platform #0: CL_INVALID_DEVICE (-33) (-33)
No OpenCL devices found
Device 0: GeForce GTX 1070 Ti
Loaded 2 hashes with 2 different salts to test db from test vectors
Options used: -I /home/tezos/john/JohnTheRipper/run/kernels -cl-mad-enable -DSM_MAJOR=6 -DSM_MINOR=1 -cl-nv-verbose -D__GPU__ -DDEVICE_INFO=524306 -DSIZEOF_SIZE_T=8 -DDEV_VER_MAJOR=410 -DDEV_VER_MINOR=73 -D_OPENCL_COMPILER -DHASH_LOOPS=250 -DPLAINTEXT_LENGTH=110 -DPBKDF2_64_MAX_SALT_SIZE=107 $JOHN/kernels/tezos_kernel.cl
Calculating best GWS for LWS=32; max. 100ms single kernel invocation.
Raw speed figures including buffer transfers:
xfer: 4.096us, init: 92.992us, crypt: 8x9.003ms, res xfer: 3.328us
gws:       256	   3465c/s     7096320 rounds/s  73.860ms per crypt_all()!
xfer: 7.168us, init: 156.672us, crypt: 8x8.865ms, res xfer: 5.664us
gws:       512	   7033c/s    14403584 rounds/s  72.798ms per crypt_all()!
xfer: 20.160us, init: 165.888us, crypt: 8x8.481ms, res xfer: 10.400us
gws:      1024	  14695c/s    30095360 rounds/s  69.679ms per crypt_all()!
xfer: 38.464us, init: 159.744us, crypt: 8x8.376ms, res xfer: 19.840us
gws:      2048	  29751c/s    60930048 rounds/s  68.836ms per crypt_all()!
xfer: 75.648us, init: 168.832us, crypt: 8x8.397ms, res xfer: 39.808us
gws:      4096	  59294c/s   121434112 rounds/s  69.079ms per crypt_all()+
xfer: 149.824us, init: 223.232us, crypt: 8x16.985ms, res xfer: 78.112us
gws:      8192	  58684c/s   120184832 rounds/s 139.593ms per crypt_all()
xfer: 301.056us, init: 449.536us, crypt: 8x25.936ms, res xfer: 155.648us
gws:     16384	  76785c/s   157255680 rounds/s 213.375ms per crypt_all()+
xfer: 705.440us, init: 946.176us, crypt: 8x42.946ms, res xfer: 312.064us
gws:     32768	  92622c/s   189689856 rounds/s 353.781ms per crypt_all()+
xfer: 1.600ms, init: 1.746ms, crypt: 8x74.382ms, res xfer: 833.696us
gws:     65536	 106819c/s   218765312 rounds/s 613.521ms per crypt_all()+
xfer: 2.912ms, init: 3.404ms, crypt: 8x138.670ms (exceeds 100ms)
Calculating best LWS for GWS=65536
Testing LWS=32 GWS=65536 ... 199.911ms+
Testing LWS=64 GWS=65536 ... 201.700ms
Testing LWS=128 GWS=65536 ... 201.858ms
Testing LWS=256 GWS=65536 ... 308.872ms
Calculating best GWS for LWS=32; max. 200ms single kernel invocation.
Raw speed figures including buffer transfers:
xfer: 45.664us, init: 146.432us, crypt: 8x7.315ms, res xfer: 23.840us
gws:      2432	  40436c/s    82812928 rounds/s  60.144ms per crypt_all()!
xfer: 89.568us, init: 162.816us, crypt: 8x8.140ms, res xfer: 46.592us
gws:      4864	  72609c/s   148703232 rounds/s  66.988ms per crypt_all()+
xfer: 180us, init: 248.832us, crypt: 8x15.796ms, res xfer: 92.512us
gws:      9728	  74874c/s   153341952 rounds/s 129.924ms per crypt_all()+
xfer: 456.544us, init: 495.616us, crypt: 8x21.898ms, res xfer: 186.272us
gws:     19456	 107771c/s   220715008 rounds/s 180.530ms per crypt_all()+
xfer: 991.584us, init: 980.992us, crypt: 8x45.005ms, res xfer: 372.096us
gws:     38912	 104874c/s   214781952 rounds/s 371.032ms per crypt_all()
xfer: 2.373ms, init: 1.949ms, crypt: 8x81.987ms, res xfer: 1.041ms
gws:     77824	 114953c/s   235423744 rounds/s 677.006ms per crypt_all()+
xfer: 3.576ms, init: 4.077ms, crypt: 8x162.336ms, res xfer: 1.776ms
gws:    155648	 116216c/s   238010368 rounds/s    1.339s per crypt_all()+
xfer: 7.518ms, init: 7.971ms, crypt: 8x314.721ms (exceeds 200ms)
Local worksize (LWS) 32, global worksize (GWS) 155648
Press 'q' or Ctrl-C to abort, almost any other key for status
1 0g 0:00:00:16  0g/s 87224p/s 87224c/s 87224C/s GPU:64°C util:100% fan:0% borro16..ms2788
1 0g 0:00:00:24  0g/s 90417p/s 90417c/s 90417C/s GPU:62°C fan:8% putml..pluttos
1 0g 0:00:00:29  0g/s 90803p/s 90803c/s 90803C/s GPU:63°C fan:10% 2PGR..patarix
Waiting for 1 child to terminate
Session aborted

**Comment by magnumripper** [comment]
What if you just try this hack?
```diff
diff --git a/src/opencl_common.c b/src/opencl_common.c
index 3d0854194..c8d7a953b 100644
--- a/src/opencl_common.c
+++ b/src/opencl_common.c
@@ -444,6 +444,8 @@ static void get_opencl_environment()
 	cl_uint num_platforms, device_pos = 0;
 	int ret, i;
 
+	sleep(NODE);
+
 	/* Find OpenCL enabled devices. We ignore error here, in case
 	 * there is no platform and we'd like to run a non-OpenCL format. */
 	ret = clGetPlatformIDs(MAX_PLATFORMS, platform_list, &num_platforms);
```

**Comment by solardiz** [comment]
One thing I would try is doing a non-MPI build. Personally, I'd never even consider using MPI on a single machine, except possibly as a stress-test trying to trigger bugs. I doubt what we're seeing here is caused by MPI, but that's one of the first things to eliminate from the picture, especially as it's totally unneeded anyway.

**Comment by solardiz** [comment]
Our `INSTALL-UBUNTU` talks about "optional MPI support" as if it were a general nice-to-have - but I think it is not. It's even an unjustified security risk since the MPI library may listed on a port. I think we should add some note to there that MPI should only be compiled in if the build is for a multi-machine cluster with shared networked storage, and that merely having it compiled in (even if unused) may have security and reliability drawbacks.

**Comment by magnumripper** [comment]
> One thing I would try is doing a non-MPI build

While the OP was with MPI, he tried non-MPI in https://github.com/magnumripper/JohnTheRipper/issues/3450#issuecomment-441486800 with same results. I agree about clarifying the docs though, I'll do so.

**Comment by madtomic** [comment]
I am testing on newer release of Ubuntu and I am having issue building with OpenCL support. please see #3492 

As for this issue with Nvidia cards, in the end, I am only able to use separate terminal to achieve the desire setup. no hacks thus far let me run john fork across devices.

**Comment by magnumripper** [comment]
The thing is you are the only one I've heard of having this problem. Can you run hashcat using two cards?

**Comment by madtomic** [comment]
@solardiz 

This why I am trying to use older version of Ubuntu.
> The fglrx driver would not be supported in 16.04 (XServer 1.18 is not supported).

==== If you have an AMD GPU (or want to use CPU as OpenCL device)

    - Up to Ubuntu 15.10 (<= 15.10). This will work on Intel hardware but warnings will be seen.
    sudo apt-get install ocl-icd-opencl-dev opencl-headers fglrx-dev

    - Ubuntu Xenial 16.04 LTS and above (>= 16.04)
    The fglrx driver would not be supported in 16.04 (XServer 1.18 is not supported).
    Canonical and AMD decided to use the new hybrid driver stack for 16.04. JtR OpenCL
users should stay on a supported release until the hybrid stack is available.

**Comment by claudioandre-br** [comment]
Bug closed as:

* **Low Priority** - They have a tendency to become high priority if enough of the users complain.

* **Business Priority** - Obsolete hardware and software. Hardware nobody owns, etc.

* **ExternalDependency** - We can't do much.

* **WontFix** - Covers all the reasons we chose to close the bug without taking action.
  * Can't reproduce, working as intended, obsolete.
  * Required information was not provided.
